// @canva/editing-extensions-api/dist/canva_api.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Message;
(function (Message) {
    let Topic;
    (function (Topic) {
        Topic["WAKE_UP"] = "wake_up";
        // Initial setup
        Topic["INIT_REQUEST"] = "init_request";
        Topic["INIT_RESPONSE"] = "init_response";
        // Canva requests presets
        Topic["PRESETS_REQUEST"] = "presets_request";
        Topic["PRESETS_RESPONSE"] = "presets_response";
        // A preset was selected by the user; let the app know
        Topic["PRESET_SELECTED"] = "preset_selected";
        // Sending the controls and managing their state
        Topic["CONTROLS_RENDER"] = "controls_render";
        Topic["CONTROLS_EVENT"] = "controls_event";
        Topic["THUMBNAIL_OVERRIDE"] = "thumbnail_override";
        // Sending a new image for the app to use
        Topic["IMAGE_UPDATE"] = "image_update";
        // Resizing the viewport
        Topic["VIEWPORT_RESIZE"] = "viewport_resize";
        // Sending a rendered image from the app
        Topic["SAVE_REQUEST"] = "save_request";
        Topic["SAVE_RESPONSE"] = "save_response";
        // Process an image using the server
        Topic["START_REMOTE_PROCESS"] = "start_remote_process";
        Topic["REMOTE_PROCESS_STARTED"] = "remote_process_started";
        Topic["REMOTE_PROCESS_PROGRESS"] = "remote_process_progress";
        Topic["REMOTE_PROCESS_COMPLETE"] = "remote_process_complete";
        // The app is done and wants to exit
        Topic["EXIT"] = "exit";
    })(Topic = Message.Topic || (Message.Topic = {}));
})(Message = exports.Message || (exports.Message = {}));
