// @canva/editing-extensions-api/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const canva_api_client_1 = require("./canva_api_client");
exports.CanvaApiClient = canva_api_client_1.CanvaApiClient;
// canva image
var canva_image_1 = require("./canva_image");
exports.CanvaImage = canva_image_1.CanvaImage;
// control kit types
var control_kit_1 = require("./control_kit/control_kit");
exports.ControlName = control_kit_1.ControlName;
// control kit components
var control_kit_2 = require("./control_kit/control_kit");
exports.ColorPicker = control_kit_2.ColorPicker;
exports.Page = control_kit_2.Page;
exports.Select = control_kit_2.Select;
var group_1 = require("./control_kit/group");
exports.Group = group_1.Group;
var slider_1 = require("./control_kit/slider");
exports.Slider = slider_1.Slider;
var button_1 = require("./control_kit/button");
exports.Button = button_1.Button;
var checkbox_1 = require("./control_kit/checkbox");
exports.Checkbox = checkbox_1.Checkbox;
var paragraph_1 = require("./control_kit/paragraph");
exports.Paragraph = paragraph_1.Paragraph;
var text_input_1 = require("./control_kit/text_input");
exports.TextInput = text_input_1.TextInput;
var radio_group_1 = require("./control_kit/radio-group");
exports.RadioGroup = radio_group_1.RadioGroup;
var preview_overlay_1 = require("./control_kit/preview_overlay");
exports.PreviewOverlay = preview_overlay_1.PreviewOverlay;
var thumbnail_list_1 = require("./control_kit/thumbnail_list");
exports.ThumbnailId = thumbnail_list_1.ThumbnailId;
exports.ThumbnailList = thumbnail_list_1.ThumbnailList;
// errors
var errors_1 = require("./errors");
exports.UnreachableError = errors_1.UnreachableError;
exports.UnimplementedError = errors_1.UnimplementedError;
var canva_api_1 = require("./canva_api");
exports.Message = canva_api_1.Message;
var mailbox_1 = require("./mailbox");
exports.Envelope = mailbox_1.Envelope;
exports.Mailbox = mailbox_1.Mailbox;
var event_handler_1 = require("./event_handler");
exports.EventHandler = event_handler_1.EventHandler;
// TODO split out into `@canva/editing-extensions-host`
var host_bus_1 = require("./host/host_bus");
exports.HostBus = host_bus_1.HostBus;
var iframe_connection_1 = require("./host/iframe_connection");
exports.connect = iframe_connection_1.connect;
const canva_image_2 = require("./canva_image");
const preview_overlay_2 = require("./control_kit/preview_overlay");
const thumbnail_list_2 = require("./control_kit/thumbnail_list");
const control_kit_3 = require("./control_kit/control_kit");
if (window.__exposeCanvaOnWindow) {
    window.canva = {
        CanvaApiClient: canva_api_client_1.CanvaApiClient,
        CanvaImage: canva_image_2.CanvaImage,
        Button: control_kit_3.Button,
        Checkbox: control_kit_3.Checkbox,
        ColorPicker: control_kit_3.ColorPicker,
        Group: control_kit_3.Group,
        Page: control_kit_3.Page,
        Paragraph: control_kit_3.Paragraph,
        PreviewOverlay: preview_overlay_2.PreviewOverlay,
        Select: control_kit_3.Select,
        Slider: control_kit_3.Slider,
        TextInput: control_kit_3.TextInput,
        ThumbnailList: thumbnail_list_2.ThumbnailList,
    };
}
